massgov/rmvwaittime
# DP-247 build rmv wait time widget

This PR creates the RMV wait time widget using static component which gets populated with dynamic data from a MassDOT feed every minute.

Ticket: https://jira.state.ma.us/browse/DP-247

### Options to build locally:
1. clone repo
2. pulling branch
3. from repo root, run `npm install`
4. to test js, run: `npm run test-js`
5. to serve app, run: `gulp serve`

### To test end to end integration:
1. Browse to: https://pilot.mass.gov/southbridge-rmv-service-center
- [ ] check viewports to ensure all widget content is always visible
2. Open your develop tools, js console tab
3. Confirm that every minute, you see a new console.info with exact feed wait times and promised wait times

### Deeper functional tests:
1. Browse to: https://massgov.github.io/rmvwaittime/?town=Boston
2. Open your develop tools, js console tab
3. Confirm that every minute, you see a new console.info with exact feed wait times and promised wait times
4. Confirm that the footer "Updated at" timestamp updates correctly
5. Test error handling for no town parameter passed
- [ ] remove the ?town=Boston from the URL
- [ ] confirm that "Wait time unavailable" renders, helpful error thrown (passed to google analytics, pinged in slack every afternoon when ga sends a report)
6. Test error handling for incorrect town parameter passed
- [ ] append ?town=foo to URL
- [ ] confirm that "Wait time unavailable" renders, helpful error thrown (passed to google analytics, pinged in slack every afternoon when ga sends a report)
- [ ] script will not autoupdate
- [ ] script will not autoupdate
7. Test that town name can be case insensitive even though data feed is case sensitive
- [ ] append ?town=plymouth to URL
- [ ] confirm that wait time appears
8. When applicable, confirm that:
- [ ] When the wait time is 0, "No wait time" appears.
- [ ] If the hours/minutes = 1, we use singular (e.g. "1 hour" or "1 minute") but if it equals more than 1, we use plural (e.g., "2 hours" or "2 minutes")
- [ ] If the wait time is less than 1 minute, "Less than a minute" appears.
- [ ] If the wait time is less than 1 hour, round up to the minute.
- [ ] Always spell out words when needed: hour/s, minute/s.
- [ ] If the wait time is more than 1 hour, round up to the nearest 15 minutes.
- [ ] When the location is closed, "Closed" appears as the wait time of each service.
- [ ] If the script runs successfully at least once and subsequently fails, rather than show "Wait time unavailable", leave the last known good times (and their corresponding update time) on the page (and the script does not attempt to update anymore).

### Screenshots

Rendering on pilot.mass.gov (with console info from past minutes):
![image](https://cloud.githubusercontent.com/assets/3279883/22379041/b51a5826-e485-11e6-8e47-5b7fa9d21d55.png)



Link to screenshot of latest successful build in CI (with passing unit tests):
![image](https://cloud.githubusercontent.com/assets/3279883/22378476/b6d5a78a-e483-11e6-8ab8-44e0c5b3e73c.png)

### @TODO 
- would like more test coverage
- would like opinion re: unit testing private functions in revealing module pattern
